### DISCUSSION TITLE
plugin-installation in docker

### DISCUSSION THREAD

**Discussion by pvdh** [verdaccio/verdaccio#3466] [Q&A] [answered]
I'm basing on verdaccio/verdaccio:5 and try to include both  verdaccio-openid-connect@latest and @wunderwerk/verdaccio-static-access-token-middleware-plugin

verdaccio-openid-connect loads without issues, so I don't think I'm completely mistaken.

However, @wunderwerk/verdaccio-static-access-token-middleware-plugin never is loaded.

```
verdaccio  |  error--- plugin not found. try npm install @wunderwerk/verdaccio-static-access-token-middleware-plugin
verdaccio  | (node:8) UnhandledPromiseRejectionWarning: Error:
verdaccio  |       @wunderwerk/verdaccio-static-access-token-middleware-plugin plugin not found. try "npm install @wunderwerk/verdaccio-static-access-token-middleware-plugin"
verdaccio  |     at /opt/verdaccio/build/lib/plugin-loader.js:146:13
verdaccio  |     at Array.map (<anonymous>)
verdaccio  |     at loadPlugin (/opt/verdaccio/build/lib/plugin-loader.js:73:37)
verdaccio  |     at defineAPI (/opt/verdaccio/build/api/index.js:82:45)
verdaccio  |     at _default (/opt/verdaccio/build/api/index.js:138:10)
verdaccio  |     at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

I follow these directions, my Dockerfile is

```
# Docker multi-stage build - https://docs.docker.com/develop/develop-images/multistage-build/
# Use an alpine node image to install the plugin
FROM node:lts-alpine as builder

RUN mkdir -p /verdaccio/plugins \
  && cd /verdaccio/plugins \
  && npm install --global-style --no-bin-links --omit=optional verdaccio-openid-connect@latest \
  && npm install  --global-style --no-bin-links --omit=optional @wunderwerk/verdaccio-static-access-token-middleware-plugin

FROM verdaccio/verdaccio:5
#ENV VERDACCIO_USER_ID=10001
# copy your modified config.yaml into the image
#ADD docker.yaml /verdaccio/conf/config.yaml

COPY --chown=$VERDACCIO_USER_UID:root --from=builder \
  /verdaccio/plugins/node_modules/verdaccio-openid-connect \
  /verdaccio/plugins/verdaccio-openid-connect
COPY --chown=$VERDACCIO_USER_UID:root --from=builder \
  /verdaccio/plugins/node_modules/@wunderwerk/verdaccio-static-access-token-middleware-plugin \
  /verdaccio/plugins/@wunderwerk/verdaccio-static-access-token-middleware-plugin
COPY --chown=$VERDACCIO_USER_UID:root --from=builder \
  /verdaccio/plugins/node_modules/@wunderwerk/verdaccio-static-access-token-middleware-plugin \
  /verdaccio/plugins/node_modules/@wunderwerk/verdaccio-static-access-token-middleware-plugin
COPY  --chown=$VERDACCIO_USER_UID:root --from=builder \
   /verdaccio/plugins/node_modules/@wunderwerk/verdaccio-static-access-token-middleware-plugin \
   /usr/local/lib/node_modules/@wunderwerk/verdaccio-static-access-token-middleware-plugin
```

I started to copy the plugin to lots of folders that sound like they might work, but it did not help.

I looked at /opt/verdaccio/build/lib/plugin-loader.js, which has some special condition for scoped packages. But I'm lacking the expertise to really get it.

I try running a shell inside the container
`docker run -v /srv/verdaccio/config/:/verdaccio/config -it verdaccio-verdaccio /bin/sh`

then
`npm -g install @wunderwerk/verdaccio-static-access-token-middleware-plugin`

but it can't write to /usr/local/lib/node_modules/. Still, that's why I tried installing to that dirctory in my Dockerfile.

I can install to /opt/verdaccio/node_modules and try to start verdaccio inside the container

`node -r ./.pnp.js $VERDACCIO_APPDIR/bin/verdaccio --config /verdaccio/conf/config.yaml --listen $VERDACCIO_PROTOCOL://0.0.0.0:$VERDACCIO_PORT
`
It still tells me
` error--- plugin not found. try npm install @wunderwerk/verdaccio-static-access-token-middleware-plugin`
which is exactly that I did.

I removed the configuration section regarding verdaccio-openid-connect and active directoy from my config, to confirm there is no conflict of the plugins.

And now I'm really out of ideas.

Any help would be greatly appreciated.

Best regards
Patrick

**Comment by juanpicado** [comment, marked as answer]
Ok I can reproduce, I'll check over the next days :) I'll keep you posted, feel free to ping me if take more than 1 week (I'm mostly curious why, this should works just don't want to forget this topic).

```
 warn --- config file  - /verdaccio/conf/config.yaml
 info --- plugin successfully loaded: verdaccio-auth-memory
 info --- plugin successfully loaded: verdaccio-audit
 error--- plugin not found. try npm install verdaccio-static-token
(node:8) UnhandledPromiseRejectionWarning: Error: 
      verdaccio-static-token plugin not found. try "npm install verdaccio-static-token"
    at /opt/verdaccio/build/lib/plugin-loader.js:146:13
    at Array.map (<anonymous>)
    at loadPlugin (/opt/verdaccio/build/lib/plugin-loader.js:73:37)
    at defineAPI (/opt/verdaccio/build/api/index.js:82:45)
    at _default (/opt/verdaccio/build/api/index.js:138:10)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:8) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:8) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Reply by juanpicado** [reply]
Btw, the plugin `@wunderwerk/verdaccio-static-access-token-middleware-plugin` relies on packages for the v6 which are in alpha

```
"devDependencies": {
    "@types/debug": "^4.1.7",
    "@types/express": "^4.17.14",
    "@types/node": "^18.8.2",
    "@typescript-eslint/eslint-plugin": "^5.39.0",
    "@typescript-eslint/parser": "^5.39.0",
    "@verdaccio/config": "6.0.0-6-next.48",
    "@verdaccio/logger": "6.0.0-6-next.16",
    "@verdaccio/types": "^10.7.0",
    "@wunderwerk/eslint-config": "^1.0.0",
    "eslint": "^8.24.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jsdoc": "^39.3.6",
    "prettier": "^2.7.1",
    "typescript": "^4.8.4"
  },
```

This explains why does not work on v5.

**Reply by pvdh** [reply]
I should have spotted that dependency, sorry for the fuss.

**Reply by juanpicado** [reply]
No problem you actually gave me good reasons to double check the image

**Comment by juanpicado** [comment]
I don´t there is an "easy solution" with the current image, so I decided to refactor it, I´d like this to be a seamless process, no hacks, no weird setups. so keep posted, probably next minor would contain a big update on the docker image

**Comment by shakeyourbunny** [comment]
Same issue with the verdaccio-activedirectory plugin in docker and verdaccio/verdaccio:5 image.

**Comment by juanpicado** [comment]
I've updated the docker image

https://github.com/verdaccio/verdaccio/releases/tag/v5.16.0

```
docker run -it --rm --name verdaccio -p 4873:4873 verdaccio/verdaccio:5.16.3
```

and docs are better, hope it helps https://verdaccio.org/docs/docker#plugins

and  examples https://github.com/verdaccio/verdaccio/tree/master/docker-examples/v5/plugins
